File: 2466-count-ways-to-build-good-strings/2466-count-ways-to-build-good-strings.cpp
class Solution {
public:
    int mod = 1000000007;
    int countGoodStrings(int low, int high, int zero, int one) {
        vector<int> dp(high+1, -1);
        int ans = 0;
        for(int i=low; i<=high; i++){
            ans = ((ans%mod) + (solve(i, zero, one, dp)%mod))%mod;
        }
        return ans;
    }
    int solve(int ind, int zero, int one, vector<int>&dp){
        if(ind == 0) return 1;
        if(ind < 0) return 0;
        if(dp[ind] != -1) return dp[ind]%mod;
        int forZero = solve(ind-zero, zero, one, dp);
        int forOne = solve(ind-one, zero, one, dp);
        return dp[ind] = (forZero+forOne)%mod;
    }
};

File: 2466-count-ways-to-build-good-strings/2466-count-ways-to-build-good-strings.py
class Solution:
    def countGoodStrings(self, low, high, zero, one):
        self.mod = 1000000007
        dp = [-1] * (high + 1)
        ans = 0
        for i in range(low, high + 1):
            ans = (ans + self.solve(i, zero, one, dp)) % self.mod
        return ans
    def solve(self, ind, zero, one, dp):
        if ind == 0: return 1
        if ind < 0: return 0
        if dp[ind] != -1: return dp[ind] % self.mod
        forZero = self.solve(ind - zero, zero, one, dp)
        forOne = self.solve(ind - one, zero, one, dp)
        dp[ind] = (forZero + forOne) % self.mod
        return dp[ind]